zebra-network/src/peer_set/set.rs

```diff
@@ -447,14 +474,16 @@ where
     /// Adds a busy service to the unready list,
     /// and adds a cancel handle for the service's current request.
     fn push_unready(&mut self, key: D::Key, svc: D::Service) {
-        let (tx, rx) = oneshot::channel();
-        self.cancel_handles.insert(key, tx);
-        self.unready_services.push(UnreadyService {
-            key: Some(key),
-            service: Some(svc),
-            cancel: rx,
-            _req: PhantomData,
-        });
+        if svc.version() >= self.minimum_peer_version.current() {
```

Comment:
Possible Bug: The request has already been sent to the peer before this method is called. So the peer is in a half-disconnected state: it has a pending request future, but it has been dropped by the peer set.

Can we revert this change, and drop unready peers in `poll_unready` after their requests have completed?

The request timeout is 20 seconds, and using a peer that has just become outdated should be ok. Particularly for a single request.

Reply:
I refactored it so that it is still inserted but cancelled immediately if the peer is outdated. Is this better, or should I revert the whole change?

Reply:
"No new requests get sent to outdated peers" seems like a nice guarantee. We should document it somewhere!

It feels a bit weird to always return an error for requests to outdated peers. But it might be the right thing to do, if network upgrade activation triggers a request that only works with the new network protocol. Let's document that reasoning somewhere.

If we want to cancel it immediately, we should check that the cancel handle gets priority over any other futures in its `select`. And I guess we'll need a test for that.

Reply:
The cancel handle is checked first in `UnreadyService::poll`, so I think we're good here.